operator/controllers/resources/role_controller_test.go

```diff
@@ -33,57 +39,189 @@ import (
-
-	require.NoError(t, teleportServer.Start())
-
-	tClient := clientForTeleport(t, teleportServer, operatorName)
-	k8sClient := startKubernetesOperator(t, tClient)
-
-	ns := createNamespaceForTest(t, k8sClient)
+	ctx, tClient, k8sClient, ns := setupKubernetesAndTeleport(t)
 	roleName := validRandomResourceName("role-")
 
-	// The role is created in K8S
+	// End of setup, we create the role in Kubernetes
 	k8sCreateDummyRole(ctx, t, k8sClient, ns.Name, roleName)
 
+	// We wait for the role to be created in Teleport
 	fastEventually(t, func() bool {
 		tRole, err := tClient.GetRole(ctx, roleName)
 		if trace.IsNotFound(err) {
 			return false
 		}
 		require.NoError(t, err)
 
+		// Role should have the same name, and have the Kubernetes origin label
 		require.Equal(t, tRole.GetName(), roleName)
-
 		require.Contains(t, tRole.GetMetadata().Labels, types.OriginLabel)
 		require.Equal(t, tRole.GetMetadata().Labels[types.OriginLabel], types.OriginKubernetes)
 
 		return true
 	})
 
-	// The role is deleted in K8S
+	// Cleanup and test, we delete the role in Kubernetes```
 	k8sDeleteRole(ctx, t, k8sClient, roleName, ns.Name)
 
+	// We wait for the role to be deleted in Teleport
 	fastEventually(t, func() bool {
 		_, err := tClient.GetRole(ctx, roleName)
 		return trace.IsNotFound(err)
 	})
 }
 
-func TestRoleDeletionDrift(t *testing.T) {
-	ctx := context.Background()
+func TestRoleCreationFromYAML(t *testing.T) {
+	ctx, tClient, k8sClient, ns := setupKubernetesAndTeleport(t)
+	tests := []struct {
+		name         string
+		roleSpecYAML []byte
+		shouldFail   bool
+		expectedSpec *types.RoleSpecV5
+	}{
+		{
+			"Valid login list",
+			[]byte(`
+allow:
+  logins:
+  - ubuntu
+  - root
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					Logins: []string{"ubuntu", "root"},
+				},
+			},
+		},
+		{
+			"Valid node_labels wildcard (list version)",
+			[]byte(`
+allow:
+  node_labels:
+    '*': ['*']
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					NodeLabels: map[string]apiutils.Strings{
+						"*": {"*"},
+					},
+				},
+			},
+		},
+		{
+			"Valid node_labels wildcard (string version)",
+			[]byte(`
+allow:
+  node_labels:
+    '*': '*'
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					NodeLabels: map[string]apiutils.Strings{
+						"*": {"*"},
+					},
+				},
+			},
+		},
+		{
+			"Invalid node_labels (label value is integer)",
+			[]byte(`
+allow:
+  node_labels:
+    'foo': 1
+`),
+			true,
+			nil,
+		},
+		{
+			"Invalid node_labels (label value is object)",
+			[]byte(`
+allow:
+  node_labels:
+    'foo':
+      'bar': 'baz'
+    'logins':
+      - 'ubuntu'
+`),
+			true,
+			nil,
+		},
+	}
 
-	teleportServer, operatorName := defaultTeleportServiceConfig(t)
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Creating the Kubernetes resource. We are using an untyped client to be able to create invalid resources.
```

Comment:
I think we can mark the inner `t *testing.T` test as Parallel 
Each test has its own role name, so it should be safe